File: java/src/main/lowestCommonAncestor/recursive.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if((p.val -root.val) * (q.val - root.val) <= 0){
            return root;
        }
        else if( p.val > root.val){
            return lowestCommonAncestor(root.right, p, q);
        }else {
            return lowestCommonAncestor(root.left, p, q);
        }
    }
}

File: python/235LowestCommonAncestorofaBinarySearcgTree/recursive.py
class Solution:
    def lowestCommonAncestor(self, root:'TreeNode', p: 'TreeNode', q:
            'TreeNode') -> 'TreeNode':
        if not root:
            return
        return self.lca(root, p, q)
    
    def lca(self, root, p, q):
        if not root:
            return 
        if root==p or root==q:
            return root
        left = self.lca(root.left, p,q)
        right = self.lca(root.right, p, q)
        if left and right:
            return root
        if not left:
            return right
        if not right:
            return left
